mm2src/kdf_walletconnect/src/client/stream.rs

```diff
@@ -0,0 +1,293 @@
+use crate::client::{create_request, HttpRequest, InboundRequest, MessageIdGenerator, OutboundRequest, ResponseFuture};
+use crate::error::{ClientError, CloseReason, WebsocketClientError};
+
+use futures_util::{stream::FusedStream, Stream};
+use futures_util::{SinkExt, StreamExt};
+use mm2_err_handle::prelude::{MapToMmResult, MmResult};
+use relay_rpc::domain::MessageId;
+use relay_rpc::rpc::{self, Params, Payload, Response, ServiceRequest, Subscription};
+use std::{collections::{hash_map::Entry, HashMap},
+          pin::Pin,
+          task::{Context, Poll}};
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::sync::oneshot;
+use tokio_tungstenite_wasm::{connect, CloseFrame, Message, WebSocketStream};
+
+/// Possible events produced by the [`ClientStream`].
+///
+/// The events are produced by polling [`ClientStream`] in a loop.
+#[derive(Debug)]
+pub enum StreamEvent {
+    /// Inbound request for receiving a subscription message.
+    ///
+    /// Currently, [`Subscription`] is the only request that the Relay sends to
+    /// the clients.
+    InboundSubscriptionRequest(InboundRequest<Subscription>),
```

Comment:
nit: I would go for calling this just plain `InboundRequest` to be general and not mention specific names in this enum. and move the possible request types (currently only this one) into another enum.